KOGITO-4539: [DMN Designer] DMN takes too long to open models with too many nodes

karreiro opened the pull request:
**JIRA**: [KOGITO-4539](https://issues.redhat.com/browse/KOGITO-4539): [DMN Designer] DMN takes too long to open models with too many nodes

![2021-02-16 13 10 21](https://user-images.githubusercontent.com/1079279/108061341-6fcbeb00-7058-11eb-9478-7447cde39e82.gif)

**Business Central**:
- [WAR before this PR](https://www.dropbox.com/s/qd72jj9052ycdwk/KOGITO-4539-business-central-before.war?dl=0)
- [WAR after this PR (v1)](https://www.dropbox.com/s/jshovaxsvu0e1zf/KOGITO-4539-business-central-afterr.war?dl=0)
- [WAR after this PR (v2)](https://www.dropbox.com/s/xjwc51msbx0x4sc/KOGITO-4539-business-central-after-v2.war?dl=0)
- [WAR after this PR (v3)](https://www.dropbox.com/s/sbm3xet3yeh6qvh/KOGITO-4539-business-central-after-v3.war?dl=0)

**VS Code**:
- [plugin before this PR](https://www.dropbox.com/s/1algszsvrj8mput/KOGITO-4539-before.vsix?dl=0)
- [plugin after this PR (v1)](https://www.dropbox.com/s/mf8g215kboa3mvq/KOGITO-4539-after.vsix?dl=0)
- [plugin after this PR (v2)](https://www.dropbox.com/s/b4vf5mmnycce2t5/KOGITO-4539-after-v2.vsix?dl=0)
- [plugin after this PR (v3)](https://www.dropbox.com/s/vgjamj92iv3vnav/KOGITO-4539-after-v3.vsix?dl=0)

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* Retest PR: <b>jenkins retest this</b>
* A full downstream build: <b>jenkins do fdb</b>
* A compile downstream build: <b>jenkins do cdb</b>
* A full production downstream build: <b>jenkins do product fdb</b>
* An upstream build: <b>jenkins do upstream</b>
</details>

karreiro commented:
Jenkins do fdb.

jomarko reviewed (changes requested):
The changes caused two decision navigator icons on left side panel and broke our selenium tests.

![Screenshot from 2021-02-24 11-11-33](https://user-images.githubusercontent.com/8044780/108985327-36ffc780-7691-11eb-80e5-733dc040d70a.png)

karreiro commented:
Thank you, @jomarko. I just updated the PR with the fix and with the new artifacts in the description.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorDock.java:
```diff
@@ -70,25 +70,30 @@ public void destroy() {
 
     @Override
     public void open() {
+        // Empty.
```
What does it mean `// empty` please, I think this comment does not explain enough why the implementation does nothing.

jomarko reviewed (commented):
Code review done, tasks to be performed during manual review as we touched `enableRefreshHandlers`, verify decision navigator is updated
- [ ] after closing
- [ ] opening other diagram
- [ ] switching tabs
- [ ] switching DRDs
- [ ] switching expression and graph editor

karreiro commented:
Thank you for the feedback, @jomarko.

I've changed my approach a little bit to avoid too many re-renders on Decision Navigator. Now, it's covering all scenarios we're targeting.

I just updated the PR with the fix and with the new artifacts in the description.

karreiro commented:
Jenkins do fdb.

jomarko reviewed (changes requested):
We observe & handle same event twice.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorPresenter.java:
```diff
@@ -140,42 +144,46 @@ void setupView() {
-            refreshTreeView();
-            refreshComponentsView();
-        }
+        deferredRefresh();
```
Well spotted, @jomarko - thank you! I've applied the change in a new commit.

karreiro commented:
Jenkins do fdb.

karreiro commented:
Thank you, @jomarko! I've applied the change in a new commit.

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/DecisionNavigatorPresenter.java:
```diff
@@ -116,7 +116,7 @@ public void onRefreshDecisionComponents(final @Observes RefreshDecisionComponent
     }
 
     public void onElementAdded(final @Observes CanvasElementAddedEvent event) {
-        refreshComponentsView();
+        deferredRefresh();
     }
```
@karreiro could you please explain why not to remove `public void onElementAdded(final @Observes CanvasElementAddedEvent event)` completely?

if I understand correctly now we call the same `refresh()` twice.
- `DecisionNavigatorObserver#onCanvasElementAdded()`
- `DecisionNavigatorPresenter#onElementAdded()`

My expectation is `DecisionNavigatorPresenter` will not contain any `@Observes CanvasElementAddedEvent event` as we observe the same and handle in same way in the `DecisionNavigatorObserver` - which is `@ApplicationScoped`. What am I missing?

karreiro replied:
@jomarko your understanding is correct and both methods were being triggered, but due to the defer mechanism the logic weren't being executed twice. Anyhow, both observers are redundant indeed, so I've removed `DecisionNavigatorPresenter#onElementAdded` in a new commit :-) thanks!

karreiro commented:
Jenkins do fdb.

karreiro commented:
Thank you, @jomarko! I've applied the change in a new commit.

jomarko reviewed (approved):
opening took ~30 seconds with a memory settings:
```
JAVA_OPTS="-Xms512m -Xmx2G -XX:MetaspaceSize=96M -XX:MaxMetaspaceSize=1G -Djava.net.preferIPv4Stack=true"
```
All nodes were placed correctly.

Thank you @karreiro

danielzhe reviewed (approved):
Great work!!! Thank you!
